wiki/Client Security.md

```diff
@@ -0,0 +1,95 @@
+
+As described in Protocol § 1.1.5, each leaf node at index *i* is computed by the following process, which we will denote as ** GenLeaf**:
+
+1. Taking first 26 bytes of *k<sub>h</sub>*, and concatenate with 2 bytes of nonce, and 4 bytes of OTP corresponding to index *i*. All bytes are in big endian encoding.
+2. Hash the 32-byte result in (1) (e.g. SHA256 in Protocol). Denote this as EOTP.
+3. Hash the 32-byte output in (2) (e.g. SHA256 in Protocol). This is the value of a leaf node.
+
+For brevity, we refer to the time when the client needs to execute some operation (such as transfer) that requires an OTP code as **runtime**. We refer to the process when the wallet is created as **setup**.
+
+### Solutions
+
+#### I. Controlled Randomness
+
+We can alter step (1) of the **GenLeaf** process by taking first 22 bytes of *k<sub>h</sub>*, and replace the other 4 bytes with a randomly generated number between [1, µ], where µ is a difficulty parameter we choose. After EOTP and its leaf is generated, the randomly generated number is discarded. Setup would not be slowed down because sampling from [1, µ] can be done in constant time. The randomness would not affect the recoverability of the wallet because the setup can be made repeatable and deterministic by using the OTP seed as the seed for random number generators. 
+
+At runtime, the client possessing the correct OTP would still have to iterate through [1, µ] and use the altered step (1) with the steps (2) and (3) above until it can produce a hash value equals to the value in the leaf node. In comparison, an attacker without the correct OTP would have to enumerate permutations of [1, µ] and [1, M). In other words, the attacker would have to perform M times more operations than the user.
+
+If we choose µ=10<sup>6</sup>, the attacker would need compute 10<sup>12</sup> hashes to brute-force both the randomly generated number and the OTP. We know µ=10<sup>6</sup> is a practical number for runtime, since in ONE Wallet v0.1 we perform this number of hashes every time we generate a new wallet, and it is typically completed within less than a few seconds in the client's browser. 
+
+Using the [hashcat benchmark](https://github.com/siseci/hashcat-benchmark-comparison/blob/45a27b32a2f24d317cc29741d64fc739f3a30cb5/1x%20Gtx%201080%20TI%20with%20Overclock%20Hashcat%20Benchmark) for reference, a highend NVIDIA GPU would enable the attacker to complete the brute-force in roughly 10 minutes. However, if we require the user to provide two OTPs inputs, the attacker's job would become practically impossible: now he would need 19 years to complete the bruteforce using a highend NVIDIA GPU. The attacker would need to design an ASIC on par with the performance of AntMiner in Bitcoin mining (~100TH/s) to complete the bruteforce in a reasonable amount of time. At ~100TH/s, the time required is ~3 hours.
```

Comment:
By @ivan-homoliak-sutd : 

> You further mention that user has to provide 2 OTPs - I agree, more entropy is always better to protect against bruteforcing. However, the problem of using TOTPs is that the user has to wait (from 5 - 25 sec) to get the next TOTP from GA, which deteriorates the usability. So, it'd be good to consider the transferred amount and use 2 TOTPs when you are over a threshold, e.g., 100USD (Depending on the cost of the attack bruteforcing a single TOTP + controlled randomness)

Reply:
The proposal I had in mind is to generate another seed and provide another QR code for the user to scan. Afterwards, we regenerate the Merkle Tree using combined OTPs, and use a combined hashed seed (by taking half of each hashed seed). We will need to replace the acting root in the contract during the process. At run-time, the two OTPs would be simultaneously visible to the user, so the user needs not to wait.

Reply:
Ok, this sounds better, though it might be a bit unclear for the user to keep two OneWallet entries in Google Auth. She might accidentally drop one due to suspicion of duplicity.

Reply:
That shouldn't be an issue. We can set arbitrary names (issuer, account) in Google Authenticator's UI.

Reply:
Ok then, but keep it in mind that this needs to be investigated during some alpha testing / usability studies.